### PR TITLE
Add tightening and culling functions to gossip networks.

### PR DESCRIPTION
**Pull request by spencerkimball**
Rewrote the previous recently-broken mechanism for tightening gossip
networks to achieve the required maximum number of hops between any
two nodes to communicate gossiped infos.

In the new mechanism, nodes exchange not just high water timestamps,
but also the minimum number of hops each node has seen for an info
from each node in the network. New infos are received either when
the timestamp is newer or the number of hops is smaller, to
acknowledge when "tighter" connections are made.

As new infos are received, nodes will make new connections to the
most "distant" node in the reachable network, as measured by number
of hops the info(s) needed to cross the network. Distant nodes are
defined as ones where the number of hops is greater than the constant
value for MaxHops (set to 5 currently).

Each node also periodically culls the least useful outgoing connection
in order to keep the network from getting too locked into a sub-optimal
configuration.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3501)

### CODE REVIEW

**Comment by spencerkimball** [comment]
The .zip file contains two subdirectories showing the evolution of a gossip network of 25 nodes with 1 bootstrap location (node 1), with and without network tightening. Each node measure average age (AA) of infos it's received as well as the maximum number of hops (MH) and age of the sentinel (SA). A node is red if any other node is unreachable. New connections are green. The simulations run until the network is fully connected and there are no further changes to connections.

[graphs.zip](https://github.com/cockroachdb/cockroach/files/67640/graphs.zip)

File: gossip/server.go [hunk 1]
```diff
@@ -99,45 +101,49 @@ func (s *server) Gossip(argsI proto.Message) (proto.Message, error) {
 	// as a permanent peer. We always accept its input and return
 	// our delta.
 	canAccept := true
-	if !s.incoming.hasNode(args.NodeID) {
-		s.incoming.setMaxSize(s.maxPeers())
-		if !s.incoming.hasSpace() {
-			canAccept = false
+	if args.NodeID != 0 {
```
Comment (spencerkimball) [inline gossip/server.go]: The changes in this file are actually really minimal–it's basically rearranging things so we no longer trade gossip if the connection will be forwarded (doing so made computation of hops between any two nodes imaccurate).

**Comment by tschottdorf** [comment]
LGTM. Thanks for the detailed comment and graphs.

While you're here, the way the error returned from `(*infoStore).addInfo` are used to count how many infos in the delta were not fresh (in `(*infoStore).combine`) is a little sketchy. Returning `var errNotFresh = errors.New("...")` and comparing at the caller is cleaner.

---

Reviewed 12 of 13 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[gossip/client.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K60DwQxyc2KIaMDpXFo:-K60DPZyKjNZin9qjlHD:698268163) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/client.go#L163)):*
can information be lost here by simply taking the reply as the gold standard? If I'm reading this right, `reply.Nodes` is based on our knowledge at time of sending the RPC, plus the remote's knowledge at the time of replying. Seems that there's no reason we would've updated it with other remotes' info in the meantime?

---

*[gossip/gossip.proto, line 29 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K606LeFcP-fTK2SvG0q:-K606LeFcP-fTK2SvG0r:1457186017) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/gossip.proto#L29)):*
s/High/high/

---

*[gossip/gossip.proto, line 32 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K606PGT1SWZyCV0zf76:-K606PGT1SWZyCV0zf77:1347869371) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/gossip.proto#L32)):*
s/Min/min/

---

*[gossip/gossip.proto, line 69 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K606YGh0dNab08zY7z6:-K606YGh0dNab08zY7z7:1779600741) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/gossip.proto#L69)):*
s/all high water timestamps, by node,/information about other nodes, as/

---

*[gossip/info.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K607mayQAj3qCaFaQF-:-K607mayQAj3qCaFaQF0:-1560201917) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/info.go#L36)):*
in which case is `n==nil`?

---

*[gossip/infostore.go, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K608LPV5oKEykxkZSbT:-K608LPV5oKEykxkZSbU:290124) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/infostore.go#L152)):*
what's the motivation for treating something as fresh when it comes in with a lower number of hops (other than wanting to update/propagate the information about the originating node)? Wouldn't it still be a known piece of information? I'm asking mostly about the callbacks firing in that case.

---

*[gossip/server.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K60GwnV9zEen673t1b6:-K60GwnV9zEen673t1b7:627995772) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/server.go#L238)):*
add comment (I hope I got that right):
// The formula below gives the number of peers required in a network of the same size with at most `MaxHops-1` (instead of `MaxHops` to account for imperfection such as "redundant" connections between clustered nodes).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-:-K60CGluJiUFOAmcJDXz:-1114277222)*

File: gossip/gossip.go [hunk 2]
```diff
@@ -554,15 +564,16 @@ func (g *Gossip) bootstrap(stopper *stop.Stopper) {
...
 				return
 			case c := <-g.disconnected:
 				g.doDisconnected(stopper, c)
-			case <-time.After(g.jitteredInterval(checkInterval)):
-				g.doCheckNetwork(stopper)
+			case nodeID := <-g.tighten:
+				g.tightenNetwork(stopper, nodeID)
+			case <-time.After(g.jitteredInterval(cullInterval)):
+				g.cullNetwork()
 			case <-time.After(g.jitteredInterval(stallInterval)):
```
Comment (nvanbenschoten) [inline gossip/gossip.go]: I must be missing something, but for the life of me I can't figure out how these two `time.After` channels ever allowed the larger delay (`checkInterval` before, now `cullInterval`) to trigger. Based on my understanding, each iteration in the for loop would create new timers in the select block, and the timer with the shorter duration would always trigger before the other one. The duration jittering isn't strong enough to allow the larger delay to ever be smaller than the jittered shorter delay, so I can't see it ever triggering `doCheckNetwork` or `cullNetwork`. Here's a [Go playground](https://play.golang.org/p/_nPsNNSEAm) to demonstrate what I think would result, along with an [example](https://play.golang.org/p/jS-1Q0EDD-) of using a `time.Ticker` to achieve what I believe to be the desired interaction.

Comment (tschottdorf) [inline gossip/gossip.go]: when it seems like something doesn't make sense, sometimes it just doesn't.. :-)
You're absolutely right, and your ticker example is the code that should be here.
Doesn't seem like we're exercising this code path in tests, so we should change
that.

**Comment by tamird** [comment]
Reviewed 8 of 13 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[gossip/info.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K607mayQAj3qCaFaQF-:-K6UOvNIcwwpjBN9aSIE:-451794791) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/info.go#L36)):*
I think this is a consequence of the map having pointer values. @spencerkimball it is possible to generate value keys with `castvalue` (see https://github.com/gogo/protobuf/pull/121#discussion_r46022458)

---

*[gossip/simulation/gossip.go, line 211 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K6UNzCZ0Ad89ABL-Miw:-K6UNzCZ0Ad89ABL-Mix:-806880111) ([raw file](https://github.com/cockroachdb/cockroach/blob/83b061cd656ac1cac316a1d561b97b89e44c5b31/gossip/simulation/gossip.go#L211)):*
the error is now lost completely. is that intended?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501)*

File: gossip/gossip.go (hunk 2 above)
Comment (spencerkimball) [inline gossip/gossip.go]: Good catch. Adding a unittest too.

**Comment by spencerkimball** [comment]
OK, done.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[gossip/client.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K60DwQxyc2KIaMDpXFo:-K78uYyLxBGWF31knl87:296194762) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/client.go#L163)):*
Each client object is connected to a single remote gossip server. Each client maintains that server's conception of the state of other nodes in the network so that it can tailor which, if any, new information it receives should be sent to the remote server. There's no overlap between other remote gossip servers.

---

*[gossip/info.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K607mayQAj3qCaFaQF-:-K78vhtUprIY-7lBGFC3:-1031279042) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/info.go#L36)):*
n==nil when the requesting node hasn't yet "heard" of a node which originated the info, so the node isn't yet in its `nodes` map.

@tamir, I don't have any problem with the maps having pointer values. Is the castvalue stuff fully supported?

---

*[gossip/infostore.go, line 152 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K608LPV5oKEykxkZSbT:-K78wGZhVQ-4Al5eXf8S:-905840352) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/infostore.go#L152)):*
We need to update the min hops to various remote nodes in order to avoid tightening to the same remote node endlessly.

---

*[gossip/simulation/gossip.go, line 211 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K6UNzCZ0Ad89ABL-Miw:-K78zu9hxp7xNVfWq7F6:2069051620) ([raw file](https://github.com/cockroachdb/cockroach/blob/83b061cd656ac1cac316a1d561b97b89e44c5b31/gossip/simulation/gossip.go#L211)):*
Yes. It's not particularly important here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-:-K78uJjtZEzGyEfrOvlA:1336915767)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[gossip/info.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K607mayQAj3qCaFaQF-:-K79CM04JBktjV174y63:1327324132) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/info.go#L36)):*
Not sure what you mean by fully supported - it worked fine to the extent that I tested it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501)*

**Comment by spencerkimball** [comment]
Review status: 1 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[gossip/info.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K607mayQAj3qCaFaQF-:-K7AAgkfOjGeI83zKFxl:100313749) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/info.go#L36)):*
Doesn't seem like a big deal either way. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501)*

**Comment by tschottdorf** [comment]
LGTM

---

Reviewed 14 of 14 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[gossip/gossip.go, line 579 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K7B134dA-y4F2Houz4M:-K7B134eCNq1IkQ4b1oc:-2059698723) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ff5e18825c2024f304e5423925dbdb569f8e70/gossip/gossip.go#L579)):*
```
defer cullTicker.Stop()
defer stallTicker.Stop()
```

---

*[gossip/gossip_test.go, line 106 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K7B1CzdFziEep_rKjcR:-K7B1CzdFziEep_rKjcS:-1301904824) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ff5e18825c2024f304e5423925dbdb569f8e70/gossip/gossip_test.go#L106)):*
s/vv/v

---

*[gossip/info.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K607mayQAj3qCaFaQF-:-K7B3fz0riSdcdLsmOA-:-75031035) ([raw file](https://github.com/cockroachdb/cockroach/blob/6acf46484ca79d92a531e822dd5d700c5cae8c9b/gossip/info.go#L36)):*
fine with me as is.

---

*[gossip/infostore.go, line 334 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K7B4X2SyYOnTNGmQjt2:-K7B4X2SyYOnTNGmQjt3:1596171114) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ff5e18825c2024f304e5423925dbdb569f8e70/gossip/infostore.go#L334)):*
why this change?

---

*[gossip/simulation/gossip.go, line 211 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K6UNzCZ0Ad89ABL-Miw:-K7B4gTgPA312f04Bu4F:965487226) ([raw file](https://github.com/cockroachdb/cockroach/blob/83b061cd656ac1cac316a1d561b97b89e44c5b31/gossip/simulation/gossip.go#L211)):*
this is kind of a weird contract, can you (at least) add a comment?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-:-K7B5Tz2x6ZFRC3GRFta:676834233)*

**Comment by spencerkimball** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[gossip/infostore.go, line 334 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501#-K7B4X2SyYOnTNGmQjt2:-K7CviUniQRe20SV8hWZ:-1358992058) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6ff5e18825c2024f304e5423925dbdb569f8e70/gossip/infostore.go#L334)):*
Before we were tracking the total number of infos from each peer. Now we're tracking the total number of nodes reachable with minimum hops through each peer, which is a much better estimate of how "useful" a node is to avoid dropping one which is giving us a solid connection to other nodes.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3501)*

**Comment by mberhault** [comment]
It seems that this broke load-balancer based gossip discovery.
When running with --gossip=lb=...:
cockroach start --logtostderr=true --stores=ssd=data --insecure --gossip=lb=
cockroach-marc-elb-1476465324.us-east-1.elb.amazonaws.com:26257

I get:
I0104 20:30:41.650557 1587 gossip/gossip.go:736  starting client to
cockroach-marc-elb-1476465324.us-east-1.elb.amazonaws.com:26257
I0104 20:30:41.658419 1587 gossip/client.go:77  closing client to
cockroach-marc-elb-1476465324.us-east-1.elb.amazonaws.com:26257:
gossip/server.go:94: node 2: connection already closed from node 0 (
172.31.63.137:43534); ignoring gossip
I0104 20:30:41.683946 1587 server/node.go:272  store store=0:0 ([ssd]=data)
not bootstrapped
I0104 20:30:41.683973 1587 server/node.go:375  connecting to gossip network
to verify cluster ID...
I0104 20:30:42.651785 1587 gossip/gossip.go:736  starting client to
cockroach-marc-elb-1476465324.us-east-1.elb.amazonaws.com:26257
I0104 20:30:42.658931 1587 gossip/client.go:77  closing client to
cockroach-marc-elb-1476465324.us-east-1.elb.amazonaws.com:26257:
gossip/server.go:94: node 1: connection already closed from node 0 (
172.31.63.137:43536); ignoring gossip
etc....

**Comment by mberhault** [comment]
This particular log is from node 0.

**Comment by mberhault** [comment]
sorry, no it's not, it's from node 2.
node 1 had the same problem and worked when restarting it with
"--gossip=http-lb=..." which fetches the node address
from /_status/details/local as opposed to using the one passed on the
command line.

**Comment by tschottdorf** [comment]
This has been notoriously flaky because it's not in our standard CI cycle,
any chance of writing a unit test which reproduces whatever it is a lb
resolver breaks?

**Comment by spencerkimball** [comment]
@mberhault as we discussed, the `lb` resolver should be retired in favor of the `node_lookup` resolver from here on out as the TCP load balancer solution for gossip bootstrap hosts won't work for a variety of reasons going forward.

Ultimately we'll be able to retire this kludgy reliance on the outgoing port number to associate connections between Go RPC message and closed connections at the RPC subsystem layer once we move to gRPC.

**Comment by tschottdorf** [comment]
@spencerkimball is the port number reliance the reason why this is broken? I see the need to move off `lb` but I'd like to know what went wrong.

**Comment by spencerkimball** [comment]
@tschottdorf yes, that's the reason this broke. The TCP load balancer solution will never work with our reliance on outgoing port numbers in order to determine unique peers. It worked before, so far as it ever did, because we were allowing multiple connections from the same remote ip / outgoing port combo.

**Pull request merged**